proposals/NNNN-method-keypaths.md

```diff
@@ -0,0 +1,667 @@
+
+This idiom could become commonplace for APIs that require a method reference to be weakly bound. (It could even
+generalize to a callable `WeakUnappliedMethod` type, given some additional language work to support more finely typed
+`@dynamicCallable` and tuple splatting.)
+
+
+## Detailed design
+
+In general, for a type `T`and a method `m`, if and only if `T.m` evaluates to an unbound method of type
+`(T) -> (A) -> B`, then:
+
+- `\T.m` will evaluate to a read-only key path of type `KeyPath<T, (A) -> B>`, and
+- given `var x: T`, the expression `x[keyPath: \.m]` will be equivalent to the expression `x.m`.
+
+Note that this phrasing aims to extend the current behavior of key paths according to existing precedent set by unbound
+methods, and does not seek to make any novel design decisions or solve additional problems beyond that. In particular,
+just as with unbound methods:
+
+- You may provide argument names to disambiguate methods with the same name, and may omit argument names if the method
+    reference is unambiguous without them:
+
+    ```swift
+    \Array<String>.reversed            // OK (reversed() has no args)
+    \Array<String>.joined(separator:)  // OK
+    \Array<String>.joined              // Also OK (no ambiguity)
+
+    \Array<Int>.prefix          // compile error: ambiguous
+    \Array<Int>.prefix(upTo:)   // OK
+    \Array<Int>.prefix(while:)  // OK
+    ```
+
+- You may disambiguate methods overloaded by type using a contextual type:
+
+    ```swift
+    struct S {
+        func f(arg: Int) { … }
+        func f(arg: String) { … }
+    }
+
+    \S.f                               // compile error: ambiguous
+    \S.f(arg:)                         // compile error: still ambiguous
+    \S.f as KeyPath<S, (Int) -> Void>  // OK
+    ```
+
+- Key paths cannot reference methods rooted in a generic type or provided by a conditional conformance unless the root
+    type is unambiguous:
+
+    ```swift
+    \Array.joined            // OK because joined is on an extension constrained to Element == String
+    \Array.contains          // compile error: ambiguous
+    \Array<String>.contains  // OK
+    ```
+
+- The root type of a method reference cannot be a protocol with Self or associated type requirements:
+
+    ```swift
+    \Sequence.min  // compile error
+
+    // Use a generic type instead:
+    func foo<T: Sequence>(…) {
+        \T.min  // OK
+    }
+    ```
+
+- In contrast to subscripts, whose arguments may be embedded in key paths, argument values / partial method application
+    are _not_ allowed in this version of the proposal (see [Future Directions](#future-directions)):
+
+    ```swift
+    \Array.joined(separator: "\n")  // syntax error
+    ```
+
+- Method key paths may not reference mutating methods:
+
+    ```swift
+    \Array.append  // compile error (how would we reference the mutated collection?)
+    ```
+
+- Method key paths may not reference type methods (i.e. static or class methods) or operators:
```

Comment:
i.e. takes a comma (i.e., comma is required here)

Reply:
Where? I don't think one is.